stellargraph/layer/knowledge_graph.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from tensorflow.keras import activations, initializers, constraints, regularizers
+from tensorflow.keras.layers import Input, Layer, Lambda, Dropout, Reshape, Embedding
+
+from ..mapper.knowledge_graph import KGTripleGenerator
+from ..core.experimental import experimental
+
+
+class ComplExScore(Layer):
+    """
+    ComplEx scoring Keras layer.
+
+    Original Paper: Complex Embeddings for Simple Link Prediction, Théo Trouillon, Johannes Welbl,
+    Sebastian Riedel, Éric Gaussier and Guillaume Bouchard, ICML
+    2016. http://jmlr.org/proceedings/papers/v48/trouillon16.pdf
+
+    This combines subject, relation and object embeddings into a score of the likelihood of the
+    link.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shape):
+        self.built = True
+
+    def call(self, inputs):
+        s_re, s_im, r_re, r_im, o_re, o_im = inputs
```

Comment:
Sorry about that - I'll make a new comment with suggestions below referring to the right bit of code

Reply:
```suggestion
        r_re, r_im, s_re, s_im, o_re, o_im = inputs
```

For me, keeping the ordering consistent with what the `inner` function expects would improve the readability - but again its a minor thing :)

Reply:
The structure is like this because `s, r, o` matches visualising this as `<node> --<edge type>-- <node>`, and it's how the paper phrases the scoring function `ɸ(s, r, o)` (sometimes at least), and seems to be a common way to phrase such knowledge graph computations (e.g. RESCAL, TransE and even NTN in the screenshot in the PR description). The `inner` function matches how the paper describes the inner products themselves:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/1203825/74617197-19477a80-5180-11ea-81cc-a339c4e72b2f.png">

I can see what you're getting at... but I don't know whether uniformity here is the right choice.

Reply:
ah I see what you mean, agreed better to be consistent with the equations in that case :)